### PR TITLE
180 dsl 2

### PR DESCRIPTION
**Pull request by giogonzo**

### CODE REVIEW

File: src/Cache.ts [hunk 1]
```diff
@@ -43,7 +43,6 @@ export class Cache<A, L, P> {
   ): BehaviorSubject<CacheValue<L, P>> => {
     if (!this.member(input)) {
       this.subjects.set(input, new BehaviorSubject(cacheValueInitial<L, P>()));
-      this.createPending(input);
```
Comment (fes300) [inline src/Cache.ts]: is this related to this pr?

File: test/CachedQuery.test.ts [hunk 2]
```diff
@@ -1,311 +1,235 @@
...
-        );
-        expect(fetchFunction.mock.calls.length).toBe(1);
-        expect(result.isRight()).toBe(true);
+        const fetch = makeTestFetch([
+          (n: number) => taskEither.of(n * 2),
+          () => fromLeft('nope')
+        ]);
+        const query = queryShallow(fetch, refetch);
+        const result = await query.run(1).run();
+        expect(result).toEqual(right(2));
```
Comment (fes300) [inline test/CachedQuery.test.ts]: this test on `result` looks redundant now that I look at it, what do you think?

File: test/CachedQuery.test.ts [hunk 3]
```diff
@@ -1,311 +1,235 @@
...
-        );
-        expect(fetchFunction.mock.calls.length).toBe(1);
-        expect(result.isLeft()).toBe(true);
+        const fetch = makeTestFetch([
+          () => fromLeft('nope'),
+          (n: number) => taskEither.of(n * 2)
+        ]);
+        const query = queryShallow(fetch, refetch);
+        const result = await query.run(1).run();
+        expect(result).toEqual(left('nope'));
```
Comment (fes300) [inline test/CachedQuery.test.ts]: this too was already tested before IMHO

File: test/CachedQuery.test.ts [hunk 4]
```diff
@@ -1,311 +1,235 @@
...
         const result2 = await query.run(1).run();
-        expect(fetchFunction.mock.calls.length).toBe(2);
-        expect(result2.isRight()).toBe(true);
+        expect(result2).toEqual(right(2));
+        fetch.assertExhausted();
```
Comment (fes300) [inline test/CachedQuery.test.ts]: by testing both of these, aren't we actually testing the util? This was wrong also with the previous implementation I think

File: test/CachedQuery.test.ts [hunk 5]
```diff
@@ -1,311 +1,235 @@
...
-        expect(result.isRight()).toBe(true);
+        const fetch = makeTestFetch([(n: number) => taskEither.of(n * 2)]);
+        const query = queryShallow(fetch, available);
+        const result = await query.run(1).run();
+        expect(result).toEqual(right(2));
         const result2 = await query.run(1).run();
-        expect(fetchFunction.mock.calls.length).toBe(1);
-        expect(result2.getOrElse('') === result.getOrElse('')).toBe(true);
+        expect(result2).toEqual(right(2));
+        fetch.assertExhausted();
```
Comment (fes300) [inline test/CachedQuery.test.ts]: In this case also `expect(result2).toEqual(right(2))` seems like not exhaustive, what is telling us that the result is not coming from the same function being called `n` times? Probably we should give it the chance to fail by giving it an array of 2 or more fetch results and checking that `itIsNotExhausted` and that the result is the first one

File: test/CachedQuery.test.ts [hunk 6]
```diff
@@ -1,311 +1,235 @@
...
-        );
-        expect(fetchFunction.mock.calls.length).toBe(1);
-        expect(result.isLeft()).toBe(true);
+        const fetch = makeTestFetch([
+          (_: number) => fromLeft('nope'),
+          (_: number) => fromLeft('nope2')
+        ]);
+        const query = queryShallow(fetch, available);
+        const result = await query.run(1).run();
+        expect(result).toEqual(left('nope'));
```
Comment (fes300) [inline test/CachedQuery.test.ts]: I realize now that all these "midway" checks that I made are totally pointless

File: test/utils/testFetch.ts [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    const te = fromNullable(results[index])
+      .map(f => f(input))
+      .foldL(() => {
+        throw new Error('test fetch out of range');
+      }, identity);
+    index += 1;
+    return te;
+  });
+
+  (fn as any).assertExhausted = () =>
```
Comment (fes300) [inline test/utils/testFetch.ts]: I feel that more often than not we don't want to check that we called the fetch function as many times as the options we gave it (this is already covered in part by `throw new Error('test fetch out of range')`) but most of the times we want to check the exact number of times it was called

Comment (giogonzo) [reply test/utils/testFetch.ts]: this is bad naming, but the check performed in the mock impl. is actually `called # times === expected.length`, so as you are suggesting. We are passing expected invocations, we want to check that each all are actually used.

File: src/Cache.ts (hunk 1 above)
Comment (giogonzo) [reply src/Cache.ts]: yes, I'll re-add it for the moment

File: test/CachedQuery.test.ts (hunk 4 above)
Comment (giogonzo) [reply test/CachedQuery.test.ts]: You mean testing `result` and `result2`? I'd still test `result2` (the final expected result) but probably not `result` as discussed above, since it is part of the prerequisite which is something tested e.g. in the very first test above

File: test/CachedQuery.test.ts (hunk 5 above)
Comment (giogonzo) [reply test/CachedQuery.test.ts]: > same function being called n times

impossible given a correct implementation of `makeTestFetch`

Comment (fes300) [reply test/CachedQuery.test.ts]: but given said implementation it is also impossible for this test to fail, am I wrong?

File: test/CachedQuery.test.ts (hunk 4 above)
Comment (fes300) [reply test/CachedQuery.test.ts]: I actually meant `expect(result2).toEqual(right(2));` and `fetch.assertExhausted();` isn't one implying the other given the util implementation?

File: test/utils/testFetch.ts (hunk 7 above)
Comment (fes300) [reply test/utils/testFetch.ts]: what I meant is more that `expected.lenght` and `results.lenght` should be different, with the current implementation if you call the fetch function a number of times > than `results.lenght` we throw (but I feel that is not what we want)

File: src/Cache.ts (hunk 1 above)
Comment (giogonzo) [reply src/Cache.ts]: done

File: test/CachedQuery.test.ts (hunk 2 above)
Comment (giogonzo) [reply test/CachedQuery.test.ts]: pushed a commit where all intermediate result checks are removed

File: test/CachedQuery.test.ts (hunk 4 above)
Comment (giogonzo) [reply test/CachedQuery.test.ts]: Yes, I think you are right. Do you see any tests where this could be false? If not, I'd remove the tests on final results too (and keep `assertExhausted` with a better name that being static is easier to copy paste in each test)

File: test/CachedQuery.test.ts (hunk 5 above)
Comment (giogonzo) [reply test/CachedQuery.test.ts]: is this basically the same question as above? `assertExhausted` implies `result2` equality and the other way around? If yes, I answered above, if not I'm not understanding

File: test/utils/testFetch.ts (hunk 7 above)
Comment (giogonzo) [reply test/utils/testFetch.ts]: I suggest we take this one offline :)

File: test/CachedQuery.test.ts (hunk 4 above)
Comment (fes300) [reply test/CachedQuery.test.ts]: I think it can be safely removed 👍

File: test/CachedQuery.test.ts (hunk 5 above)
Comment (fes300) [reply test/CachedQuery.test.ts]: it is in part, it is more linked to the last comment I made, my concern is that if we want to keep the current implementation where the `fetch` function cannot be called a number of times > `results.lenght` then our test should verify that the query `doesNotThrow`. Alternatively, we could change the util to return a `left(out of range)` to test against

Comment (fes300) [reply test/CachedQuery.test.ts]: anyway it's a "questione di lana caprina"

Comment (giogonzo) [reply test/CachedQuery.test.ts]: > our test should verify that the query doesNotThrow

ah ok got it now: I already verified that if that error is thrown, the test fails. So we should be covered

File: test/utils/testFetch.ts (hunk 7 above)
Comment (fes300) [reply test/utils/testFetch.ts]: agreed :)

File: test/CachedQuery.test.ts (hunk 4 above)
Comment (giogonzo) [reply test/CachedQuery.test.ts]: Actually, rethinking about this: by checking the value of the final result we verify that `run()` per se does not mess in any way with the return value from the test fetch function. The fact that it just returns it internally is an implementation detail. So I still think removing the intermediate results is OK (since we are already testing them in previous tests) but the new results that are checked in each test (only the final one in this case) have a reason to be tested. So I would keep the check

**Comment by fes300** [comment]
@giogonzo what do you think about merging this? I believe we have too many open PRs the risk is loosing focus

**Comment by giogonzo** [comment]
@fes300 agreed, I'd merge this
